File: packages/typescript/src/features/advanced-memory/evaluators/summarization.ts
import { requireEvaluatorSpec } from "../../../generated/spec-helpers.ts";
import { logger } from "../../../logger.ts";
import {
	type EvaluationExample,
	type Evaluator,
	type IAgentRuntime,
	type Memory,
	ModelType,
	type TextGenerationModelType,
	type UUID,
} from "../../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../../utils.ts";
import {
	initialSummarizationTemplate,
	updateSummarizationTemplate,
} from "../prompts.ts";
import type { MemoryService } from "../services/memory-service.ts";
import { logAdvancedMemoryTrajectory } from "../trajectory.ts";
import type { SummaryResult } from "../types.ts";

// Get text content from centralized specs
const spec = requireEvaluatorSpec("MEMORY_SUMMARIZATION");

function isDialogueMessage(msg: Memory): boolean {
	return (
		!(
			(msg.content?.type as string) === "action_result" &&
			(msg.metadata?.type as string) === "action_result"
		) &&
		((msg.metadata?.type as string) === "agent_response_message" ||
			(msg.metadata?.type as string) === "user_message")
	);
}

async function getDialogueMessageCount(
	runtime: IAgentRuntime,
	roomId: UUID,
): Promise<number> {
	const messages = await runtime.getMemories({
		tableName: "messages",
		roomId,
		limit: 100,
		unique: false,
	});

	let count = 0;
	for (const msg of messages) {
		if (isDialogueMessage(msg)) {
			count += 1;
		}
	}
	return count;
}

function isRecord(value: unknown): value is Record<string, unknown> {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}

function toStringArray(value: unknown): string[] {
	if (Array.isArray(value)) {
		return value
			.map((entry) => (typeof entry === "string" ? entry.trim() : ""))
			.filter(Boolean);
	}

	if (typeof value === "string") {
		return value
			.split(",")
			.map((entry) => entry.trim())
			.filter(Boolean);
	}

	if (isRecord(value) && "point" in value) {
		return toStringArray(value.point);
	}

	return [];
}

function parseSummaryResponse(text: string): SummaryResult {
	const parsed = parseKeyValueXml<Record<string, unknown>>(text);
	if (parsed) {
		const summary =
			typeof parsed.text === "string" && parsed.text.trim().length > 0
				? parsed.text.trim()
				: "Summary not available";
		const topics = toStringArray(parsed.topics);
		const keyPoints = toStringArray(parsed.keyPoints);

		if (
			summary !== "Summary not available" ||
			topics.length > 0 ||
			keyPoints.length > 0
		) {
			return { summary, topics, keyPoints };
		}
	}

	const summaryMatch = text.match(/<text>([\s\S]*?)<\/text>/);
	const topicsMatch = text.match(/<topics>([\s\S]*?)<\/topics>/);
	const keyPointsMatches = text.matchAll(/<point>([\s\S]*?)<\/point>/g);

	return {
		summary: summaryMatch ? summaryMatch[1].trim() : "Summary not available",
		topics: topicsMatch
			? topicsMatch[1]
					.split(",")
					.map((t) => t.trim())
					.filter(Boolean)
			: [],
		keyPoints: Array.from(keyPointsMatches).map((match) => match[1].trim()),
	};
}

export const summarizationEvaluator: Evaluator = {
	name: spec.name,
	description: spec.description,
	similes: spec.similes ? [...spec.similes] : [],
	alwaysRun: spec.alwaysRun ?? true,
	examples: (spec.examples ?? []) as EvaluationExample[],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
	): Promise<boolean> => {
		if (!message.content?.text) return false;

		const memoryService = runtime.getService("memory") as MemoryService | null;
		if (!memoryService) return false;

		const config = memoryService.getConfig();
		const currentDialogueCount = await getDialogueMessageCount(
			runtime,
			message.roomId,
		);
		const existingSummary = await memoryService.getCurrentSessionSummary(
			message.roomId,
		);

		if (!existingSummary) {
			return currentDialogueCount >= config.shortTermSummarizationThreshold;
		}
		const newDialogueCount =
			currentDialogueCount - existingSummary.lastMessageOffset;
		return newDialogueCount >= config.shortTermSummarizationInterval;
	},

	handler: async (runtime: IAgentRuntime, message: Memory) => {
		const memoryService = runtime.getService("memory") as MemoryService;
		if (!memoryService) {
			logger.error({ src: "evaluator:memory" }, "MemoryService not found");
			return undefined;
		}

		const config = memoryService.getConfig();
		const { roomId } = message;

		try {
			logger.info(
				{ src: "evaluator:memory" },
				`Starting summarization for room ${roomId}`,
			);

			const existingSummary =
				await memoryService.getCurrentSessionSummary(roomId);
			const lastOffset = existingSummary?.lastMessageOffset || 0;

			const allMessages = await runtime.getMemories({
				tableName: "messages",
				roomId,
				limit: 1000,
				unique: false,
			});

			const allDialogueMessages = allMessages.filter(isDialogueMessage);

			const totalDialogueCount = allDialogueMessages.length;
			const newDialogueCount = totalDialogueCount - lastOffset;

			if (newDialogueCount === 0) {
				logger.debug(
					{ src: "evaluator:memory" },
					"No new dialogue messages to summarize",
				);
				return undefined;
			}

			const maxNewMessages = config.summaryMaxNewMessages || 50;
			const messagesToProcess = Math.min(newDialogueCount, maxNewMessages);

			if (newDialogueCount > maxNewMessages) {
				logger.warn(
					{ src: "evaluator:memory" },
					`Capping new dialogue messages at ${maxNewMessages} (${newDialogueCount} available)`,
				);
			}

			const sortedDialogueMessages = allDialogueMessages.sort(
				(a, b) => (a.createdAt || 0) - (b.createdAt || 0),
			);

			const newDialogueMessages = sortedDialogueMessages.slice(
				lastOffset,
				lastOffset + messagesToProcess,
			);
			if (newDialogueMessages.length === 0) {
				logger.debug(
					{ src: "evaluator:memory" },
					"No new dialogue messages retrieved after filtering",
				);
				return undefined;
			}

			const formattedMessages = newDialogueMessages
				.map((msg) => {
					const sender =
						msg.entityId === runtime.agentId ? runtime.character.name : "User";
					return `${sender}: ${msg.content.text || "[non-text message]"}`;
				})
				.join("\n");

			const state = await runtime.composeState(message);
			let prompt: string;
			let template: string;

			if (existingSummary) {
				template = updateSummarizationTemplate;
				prompt = composePromptFromState({
					state: {
						...state,
						existingSummary: existingSummary.summary,
						existingTopics: existingSummary.topics?.join(", ") || "None",
						newMessages: formattedMessages,
					},
					template,
				});
			} else {
				const initialMessages = sortedDialogueMessages
					.map((msg) => {
						const sender =
							msg.entityId === runtime.agentId
								? runtime.character.name
								: "User";
						return `${sender}: ${msg.content.text || "[non-text message]"}`;
					})
					.join("\n");

				template = initialSummarizationTemplate;
				prompt = composePromptFromState({
					state: { ...state, recentMessages: initialMessages },
					template,
				});
			}

			const modelType = (config.summaryModelType ??
				ModelType.TEXT_NANO) as TextGenerationModelType;
			const response = await runtime.useModel(modelType, {
				prompt,
				maxTokens: config.summaryMaxTokens || 2500,
			});

			const summaryResult = parseSummaryResponse(response);
			logAdvancedMemoryTrajectory({
				runtime,
				message,
				providerName: "MEMORY_SUMMARIZATION",
				purpose: "evaluate",
				data: {
					hasExistingSummary: !!existingSummary,
					processedDialogueMessages: newDialogueMessages.length,
					totalDialogueMessages: totalDialogueCount,
					topicCount: summaryResult.topics.length,
					keyPointCount: summaryResult.keyPoints.length,
				},
				query: {
					modelType: String(modelType),
					roomId,
				},
			});

			logger.info(
				{ src: "evaluator:memory" },
				`${existingSummary ? "Updated" : "Generated"} summary: ${summaryResult.summary.substring(0, 100)}...`,
			);

			const newOffset = lastOffset + newDialogueMessages.length;
			const firstMessage = newDialogueMessages[0];
			const lastMessage = newDialogueMessages[newDialogueMessages.length - 1];

			const startTime = existingSummary
				? existingSummary.startTime
				: firstMessage?.createdAt && firstMessage.createdAt > 0
					? new Date(firstMessage.createdAt)
					: new Date();
			const endTime =
				lastMessage?.createdAt && lastMessage.createdAt > 0
					? new Date(lastMessage.createdAt)
					: new Date();

			if (existingSummary) {
				await memoryService.updateSessionSummary(existingSummary.id, roomId, {
					summary: summaryResult.summary,
					messageCount:
						existingSummary.messageCount + newDialogueMessages.length,
					lastMessageOffset: newOffset,
					endTime,
					topics: summaryResult.topics,
					metadata: { keyPoints: summaryResult.keyPoints },
				});

				logger.info(
					{ src: "evaluator:memory" },
					`Updated summary for room ${roomId}: ${newDialogueMessages.length} messages processed`,
				);
			} else {
				await memoryService.storeSessionSummary({
					agentId: runtime.agentId,
					roomId,
					entityId:
						message.entityId !== runtime.agentId ? message.entityId : undefined,
					summary: summaryResult.summary,
					messageCount: totalDialogueCount,
					lastMessageOffset: totalDialogueCount,
					startTime,
					endTime,
					topics: summaryResult.topics,
					metadata: { keyPoints: summaryResult.keyPoints },
				});

				logger.info(
					{ src: "evaluator:memory" },
					`Created summary for room ${roomId}: ${totalDialogueCount} messages summarized`,
				);
			}
		} catch (error) {
			const err = error instanceof Error ? error.message : String(error);
			logger.error(
				{ src: "evaluator:memory", err },
				"Error during summarization",
			);
		}
		return undefined;
	},
};

File: packages/rust/src/features/advanced_memory/evaluators/summarization.rs
use crate::advanced_memory::memory_service::MemoryService;
use crate::advanced_memory::prompts::{
    INITIAL_SUMMARIZATION_TEMPLATE, UPDATE_SUMMARIZATION_TEMPLATE,
};
use crate::advanced_memory::types::SummaryResult;
use crate::runtime::AgentRuntime;
use crate::types::components::{
    ActionResult, EvaluatorDefinition, EvaluatorHandler, HandlerOptions,
};
use crate::types::database::GetMemoriesParams;
use crate::types::memory::Memory;
use crate::types::state::State;
use anyhow::Result;
use std::sync::Weak;
use std::time::{SystemTime, UNIX_EPOCH};
use tracing::{debug, error, info, warn};

fn now_unix() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64
}

/// Check if a message is a dialogue message (not action_result).
/// Matches the TypeScript `isDialogueMessage` filter.
fn is_dialogue_message(msg: &Memory) -> bool {
    let content_type = msg.content.content_type.as_deref().unwrap_or("");
    let meta_type = msg
        .metadata
        .as_ref()
        .and_then(|m| match m {
            crate::types::memory::MemoryMetadata::Custom(serde_json::Value::Object(map)) => {
                map.get("type").and_then(|v| v.as_str())
            }
            _ => None,
        })
        .unwrap_or("");

    let is_action_result = content_type == "action_result" && meta_type == "action_result";
    let is_dialogue = meta_type == "agent_response_message" || meta_type == "user_message";

    !is_action_result && is_dialogue
}

/// Parse XML summary response from LLM
fn parse_summary_xml(xml: &str) -> SummaryResult {
    let summary = regex_capture(xml, r"<text>([\s\S]*?)</text>")
        .unwrap_or_else(|| "Summary not available".to_string())
        .trim()
        .to_string();

    let topics = regex_capture(xml, r"<topics>([\s\S]*?)</topics>")
        .map(|t| {
            t.split(',')
                .map(|s| s.trim().to_string())
                .filter(|s| !s.is_empty())
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();

    let key_points = regex_capture_all(xml, r"<point>([\s\S]*?)</point>");

    SummaryResult {
        summary,
        topics,
        key_points,
    }
}

fn regex_capture(text: &str, pattern: &str) -> Option<String> {
    let re = regex::Regex::new(pattern).ok()?;
    re.captures(text)
        .and_then(|c| c.get(1))
        .map(|m| m.as_str().to_string())
}

fn regex_capture_all(text: &str, pattern: &str) -> Vec<String> {
    let re = match regex::Regex::new(pattern) {
        Ok(r) => r,
        Err(_) => return vec![],
    };
    re.captures_iter(text)
        .filter_map(|c| c.get(1).map(|m| m.as_str().trim().to_string()))
        .collect()
}

/// Compose a prompt by replacing {{variable}} placeholders in a template.
fn compose_prompt(template: &str, vars: &[(&str, &str)]) -> String {
    let mut result = template.to_string();
    for (key, val) in vars {
        let placeholder = format!("{{{{{}}}}}", key);
        result = result.replace(&placeholder, val);
    }
    result
}

pub struct SummarizationEvaluator {
    runtime: Weak<AgentRuntime>,
}

impl SummarizationEvaluator {
    pub fn new(runtime: Weak<AgentRuntime>) -> Self {
        Self { runtime }
    }
}

#[async_trait::async_trait]
impl EvaluatorHandler for SummarizationEvaluator {
    fn definition(&self) -> EvaluatorDefinition {
        EvaluatorDefinition {
            name: "MEMORY_SUMMARIZATION".to_string(),
            description: "Automatically summarizes conversations to optimize context usage"
                .to_string(),
            always_run: Some(true),
            similes: Some(vec![
                "CONVERSATION_SUMMARY".to_string(),
                "CONTEXT_COMPRESSION".to_string(),
                "MEMORY_OPTIMIZATION".to_string(),
            ]),
            ..Default::default()
        }
    }

    async fn validate(&self, message: &Memory, _state: Option<&State>) -> bool {
        let text = message.content.text.as_deref().unwrap_or("");
        if text.is_empty() {
            return false;
        }

        let runtime = match self.runtime.upgrade() {
            Some(r) => r,
            None => return false,
        };

        let service_opt = runtime.get_service("memory").await;
        let service_arc = match service_opt {
            Some(s) => s,
            None => return false,
        };
        let memory_service = match service_arc.as_any().downcast_ref::<MemoryService>() {
            Some(s) => s,
            None => return false,
        };

        let config = memory_service.get_config();

        // Get dialogue message count
        let dialogue_count = get_dialogue_count(&runtime, &message.room_id).await;

        // Check if summary exists
        let existing = memory_service
            .get_current_session_summary(message.room_id.clone())
            .await
            .unwrap_or(None);

        match existing {
            None => dialogue_count >= config.short_term_summarization_threshold,
            Some(summary) => {
                let new_count = dialogue_count - summary.last_message_offset;
                new_count >= config.short_term_summarization_interval
            }
        }
    }

    async fn handle(
        &self,
        message: &Memory,
        _state: Option<&State>,
        _options: Option<&HandlerOptions>,
    ) -> Result<Option<ActionResult>> {
        let runtime = self
            .runtime
            .upgrade()
            .ok_or_else(|| anyhow::anyhow!("Runtime dropped"))?;

        let service_opt = runtime.get_service("memory").await;
        let service_arc = match service_opt {
            Some(s) => s,
            None => {
                error!("MemoryService not found");
                return Ok(None);
            }
        };
        let memory_service = match service_arc.as_any().downcast_ref::<MemoryService>() {
            Some(s) => s,
            None => {
                error!("MemoryService type mismatch");
                return Ok(None);
            }
        };

        let config = memory_service.get_config();
        let room_id = message.room_id.clone();

        info!("Starting summarization for room {}", room_id);

        let existing = memory_service
            .get_current_session_summary(room_id.clone())
            .await
            .unwrap_or(None);
        let last_offset = existing
            .as_ref()
            .map(|s| s.last_message_offset)
            .unwrap_or(0);

        // Get all messages
        let params = GetMemoriesParams {
            room_id: Some(room_id.clone()),
            table_name: "messages".to_string(),
            count: Some(1000),
            ..Default::default()
        };
        let db = runtime
            .database()
            .ok_or_else(|| anyhow::anyhow!("Database not available"))?;
        let all_messages = db.get_memories(params).await?;

        let all_dialogue: Vec<&Memory> = all_messages
            .iter()
            .filter(|m| is_dialogue_message(m))
            .collect();
        let total_dialogue_count = all_dialogue.len() as i32;
        let new_dialogue_count = total_dialogue_count - last_offset;

        if new_dialogue_count == 0 {
            debug!("No new dialogue messages to summarize");
            return Ok(None);
        }

        let max_new = config.summary_max_new_messages.unwrap_or(50);
        let messages_to_process = new_dialogue_count.min(max_new);

        if new_dialogue_count > max_new {
            warn!(
                "Capping new dialogue at {} ({} available)",
                max_new, new_dialogue_count
            );
        }

        // Sort by timestamp
        let mut sorted_dialogue: Vec<&Memory> = all_dialogue;
        sorted_dialogue.sort_by_key(|m| m.created_at.unwrap_or(0));

        let new_messages: Vec<&Memory> = sorted_dialogue
            [last_offset as usize..(last_offset + messages_to_process) as usize]
            .to_vec();

        if new_messages.is_empty() {
            debug!("No new dialogue messages after slicing");
            return Ok(None);
        }

        // Format messages
        let agent_name = {
            let char_lock = runtime.character.read().await;
            char_lock.name.clone()
        };
        let formatted = new_messages
            .iter()
            .map(|msg| {
                let sender = if msg.agent_id.as_ref() == Some(&runtime.agent_id) {
                    agent_name.as_str()
                } else {
                    "User"
                };
                format!(
                    "{}: {}",
                    sender,
                    msg.content.text.as_deref().unwrap_or("[non-text message]")
                )
            })
            .collect::<Vec<_>>()
            .join("\n");

        // Build prompt
        let prompt = if let Some(ref ex) = existing {
            let topics_str = ex
                .topics
                .as_ref()
                .map(|t| t.join(", "))
                .unwrap_or_else(|| "None".to_string());
            compose_prompt(
                UPDATE_SUMMARIZATION_TEMPLATE,
                &[
                    ("existingSummary", &ex.summary),
                    ("existingTopics", &topics_str),
                    ("newMessages", &formatted),
                ],
            )
        } else {
            // Initial summary — use ALL dialogue
            let initial_formatted = sorted_dialogue
                .iter()
                .map(|msg| {
                    let sender = if msg.agent_id.as_ref() == Some(&runtime.agent_id) {
                        agent_name.as_str()
                    } else {
                        "User"
                    };
                    format!(
                        "{}: {}",
                        sender,
                        msg.content.text.as_deref().unwrap_or("[non-text message]")
                    )
                })
                .collect::<Vec<_>>()
                .join("\n");
            compose_prompt(
                INITIAL_SUMMARIZATION_TEMPLATE,
                &[("recentMessages", &initial_formatted)],
            )
        };

        // Call LLM
        let max_tokens = config.summary_max_tokens.unwrap_or(2500);
        let params = serde_json::json!({
            "prompt": prompt,
            "maxTokens": max_tokens,
        });
        let response = runtime
            .use_model(crate::types::model::model_type::TEXT_LARGE, params)
            .await?;

        let result = parse_summary_xml(&response);
        info!(
            "{} summary: {}...",
            if existing.is_some() {
                "Updated"
            } else {
                "Generated"
            },
            &result.summary[..result.summary.len().min(100)]
        );

        let new_offset = last_offset + new_messages.len() as i32;
        let now = now_unix();

        if let Some(ex) = existing {
            let updated = crate::advanced_memory::types::SessionSummary {
                id: crate::types::primitives::UUID::new_v4(),
                agent_id: ex.agent_id,
                room_id: room_id.clone(),
                entity_id: ex.entity_id,
                summary: result.summary,
                message_count: ex.message_count + new_messages.len() as i32,
                last_message_offset: new_offset,
                start_time: ex.start_time,
                end_time: now,
                topics: Some(result.topics),
                metadata: {
                    let mut m = serde_json::Map::new();
                    m.insert(
                        "keyPoints".to_string(),
                        serde_json::to_value(&result.key_points).unwrap(),
                    );
                    Some(m)
                },
                embedding: None,
                created_at: ex.created_at,
                updated_at: now,
            };
            memory_service
                .update_session_summary(ex.id, room_id.clone(), updated)
                .await?;
            info!(
                "Updated summary for room {}: {} messages processed",
                room_id,
                new_messages.len()
            );
        } else {
            let first_ts = new_messages
                .first()
                .and_then(|m| m.created_at)
                .unwrap_or(now);
            let new_summary = crate::advanced_memory::types::SessionSummary {
                id: crate::types::primitives::UUID::new_v4(),
                agent_id: runtime.agent_id.clone(),
                room_id: room_id.clone(),
                entity_id: if message.entity_id != runtime.agent_id {
                    Some(message.entity_id.clone())
                } else {
                    None
                },
                summary: result.summary,
                message_count: total_dialogue_count,
                last_message_offset: total_dialogue_count,
                start_time: first_ts,
                end_time: now,
                topics: Some(result.topics),
                metadata: {
                    let mut m = serde_json::Map::new();
                    m.insert(
                        "keyPoints".to_string(),
                        serde_json::to_value(&result.key_points).unwrap(),
                    );
                    Some(m)
                },
                embedding: None,
                created_at: now,
                updated_at: now,
            };
            memory_service.store_session_summary(new_summary).await?;
            info!(
                "Created summary for room {}: {} messages summarized",
                room_id, total_dialogue_count
            );
        }

        Ok(None)
    }
}

/// Count dialogue messages in a room, excluding action results.
async fn get_dialogue_count(
    runtime: &AgentRuntime,
    room_id: &crate::types::primitives::UUID,
) -> i32 {
    let params = GetMemoriesParams {
        room_id: Some(room_id.clone()),
        table_name: "messages".to_string(),
        count: Some(100),
        ..Default::default()
    };

    let db = match runtime.database() {
        Some(d) => d,
        None => return 0,
    };

    let messages = db.get_memories(params).await.unwrap_or_default();
    messages.iter().filter(|m| is_dialogue_message(m)).count() as i32
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_summary_xml_valid() {
        let xml = r#"
<summary>
  <text>The user discussed their favorite coffee.</text>
  <topics>coffee, preferences, beverages</topics>
  <key_points>
    <point>User prefers dark roast</point>
    <point>User drinks 3 cups daily</point>
  </key_points>
</summary>"#;
        let result = parse_summary_xml(xml);
        assert_eq!(result.summary, "The user discussed their favorite coffee.");
        assert_eq!(result.topics, vec!["coffee", "preferences", "beverages"]);
        assert_eq!(result.key_points.len(), 2);
        assert_eq!(result.key_points[0], "User prefers dark roast");
        assert_eq!(result.key_points[1], "User drinks 3 cups daily");
    }

    #[test]
    fn parse_summary_xml_malformed() {
        let xml = "This is not XML at all";
        let result = parse_summary_xml(xml);
        assert_eq!(result.summary, "Summary not available");
        assert!(result.topics.is_empty());
        assert!(result.key_points.is_empty());
    }

    #[test]
    fn parse_summary_xml_partial_tags() {
        let xml = "<text>Just a summary</text>";
        let result = parse_summary_xml(xml);
        assert_eq!(result.summary, "Just a summary");
        assert!(result.topics.is_empty());
        assert!(result.key_points.is_empty());
    }

    #[test]
    fn parse_summary_xml_empty_topics() {
        let xml = "<text>Summary here</text><topics></topics>";
        let result = parse_summary_xml(xml);
        assert_eq!(result.summary, "Summary here");
        assert!(result.topics.is_empty());
    }

    #[test]
    fn compose_prompt_replaces_variables() {
        let tmpl = "Hello {{name}}, welcome to {{place}}.";
        let result = compose_prompt(tmpl, &[("name", "Alice"), ("place", "Wonderland")]);
        assert_eq!(result, "Hello Alice, welcome to Wonderland.");
    }

    #[test]
    fn compose_prompt_no_match() {
        let tmpl = "No placeholders here.";
        let result = compose_prompt(tmpl, &[("key", "val")]);
        assert_eq!(result, "No placeholders here.");
    }
}
